From 0d5fd2512b47e1dc25c023b7f45006b7880a9229 Mon Sep 17 00:00:00 2001
From: Paul Grandperrin <paul.grandperrin@gmail.com>
Date: Mon, 10 Aug 2026 18:49:57 +0200
Subject: [PATCH] wsi/wayland: make present-time mode changes to IMMEDIATE send
 tearing hints

If we want clients to be able to change the present mode to IMMEDIATE at
present commit time, we first need to expose all the compatible present
modes in order to allow the client to setup the swap-chain properly.

This part is inspired by what's already being done in X11's WSI.

The only real difference is that we need to call wsi_wl_surface_get_present_modes
in order to get the list of available present modes.

Then, instead of only sending the tearing hints at the swap-chain creation,
we also send them, when needed, at present commit time in
wsi_wl_swapchain_set_present_mode().
---
 src/vulkan/wsi/wsi_common_wayland.c | 176 ++++++++++++++++------------
 1 file changed, 102 insertions(+), 74 deletions(-)

diff --git a/src/vulkan/wsi/wsi_common_wayland.c b/src/vulkan/wsi/wsi_common_wayland.c
index e0cbfd03818..aafa9636707 100644
--- a/src/vulkan/wsi/wsi_common_wayland.c
+++ b/src/vulkan/wsi/wsi_common_wayland.c
@@ -1948,6 +1948,50 @@ wsi_wl_surface_check_presentation(VkIcdSurfaceBase *icd_surface,
    return VK_SUCCESS;
 }
 
+#define WSI_WL_MAX_PRESENT_MODES 3
+
+static VkResult
+wsi_wl_surface_get_present_modes(VkIcdSurfaceBase *icd_surface,
+                                 struct wsi_device *wsi_device,
+                                 uint32_t* pPresentModeCount,
+                                 VkPresentModeKHR* pPresentModes)
+{
+   VkIcdSurfaceWayland *surface = (VkIcdSurfaceWayland *)icd_surface;
+   struct wsi_wayland *wsi =
+      (struct wsi_wayland *)wsi_device->wsi[VK_ICD_WSI_PLATFORM_WAYLAND];
+
+   struct wsi_wl_display display;
+   if (wsi_wl_display_init(wsi, &display, surface->display, true,
+                           wsi_device->sw, "mesa present modes query"))
+      return VK_ERROR_SURFACE_LOST_KHR;
+
+   VkPresentModeKHR present_modes[WSI_WL_MAX_PRESENT_MODES];
+   uint32_t present_modes_count = 0;
+
+   /* The following two modes are always supported */
+   present_modes[present_modes_count++] = VK_PRESENT_MODE_MAILBOX_KHR;
+   present_modes[present_modes_count++] = VK_PRESENT_MODE_FIFO_KHR;
+
+   if (display.tearing_control_manager)
+      present_modes[present_modes_count++] = VK_PRESENT_MODE_IMMEDIATE_KHR;
+
+   assert(present_modes_count <= ARRAY_SIZE(present_modes));
+   wsi_wl_display_finish(&display);
+
+   if (pPresentModes == NULL) {
+      *pPresentModeCount = present_modes_count;
+      return VK_SUCCESS;
+   }
+
+   *pPresentModeCount = MIN2(*pPresentModeCount, present_modes_count);
+   typed_memcpy(pPresentModes, present_modes, *pPresentModeCount);
+
+   if (*pPresentModeCount < present_modes_count)
+      return VK_INCOMPLETE;
+   else
+      return VK_SUCCESS;
+}
+
 static VkResult
 wsi_wl_surface_get_capabilities2(VkIcdSurfaceBase *surface,
                                  struct wsi_device *wsi_device,
@@ -1985,46 +2029,48 @@ wsi_wl_surface_get_capabilities2(VkIcdSurfaceBase *surface,
       }
 
       case VK_STRUCTURE_TYPE_SURFACE_PRESENT_MODE_COMPATIBILITY_KHR: {
-         /* Can easily toggle between FIFO and MAILBOX on Wayland. */
+         /* All present modes are compatible with each other. */
          VkSurfacePresentModeCompatibilityKHR *compat = (void *)ext;
+
          if (compat->pPresentModes) {
             assert(present_mode);
+
+            VkPresentModeKHR present_modes[WSI_WL_MAX_PRESENT_MODES];
+            uint32_t present_mode_count = ARRAY_SIZE(present_modes);
+
+            result = wsi_wl_surface_get_present_modes(surface, wsi_device,
+                                                      &present_mode_count,
+                                                      present_modes);
+            if (result != VK_SUCCESS)
+               return result;
+
             VK_OUTARRAY_MAKE_TYPED(VkPresentModeKHR, modes, compat->pPresentModes, &compat->presentModeCount);
             /* Must always return queried present mode even when truncating. */
             vk_outarray_append_typed(VkPresentModeKHR, &modes, mode) {
                *mode = present_mode->presentMode;
             }
-            switch (present_mode->presentMode) {
-            case VK_PRESENT_MODE_MAILBOX_KHR:
-               vk_outarray_append_typed(VkPresentModeKHR, &modes, mode) {
-                  *mode = VK_PRESENT_MODE_FIFO_KHR;
+
+            for (uint32_t i = 0; i < present_mode_count; i++) {
+               if (present_modes[i] != present_mode->presentMode) {
+                  vk_outarray_append_typed(VkPresentModeKHR, &modes, mode) {
+                     *mode = present_modes[i];
+                  }
                }
-               break;
-            case VK_PRESENT_MODE_FIFO_KHR:
-               vk_outarray_append_typed(VkPresentModeKHR, &modes, mode) {
-                  *mode = VK_PRESENT_MODE_MAILBOX_KHR;
-               }
-               break;
-            default:
-               break;
             }
          } else {
             if (!present_mode) {
                wsi_common_vk_warn_once("Use of VkSurfacePresentModeCompatibilityKHR "
                                        "without a VkSurfacePresentModeKHR set. This is an "
                                        "application bug.\n");
-               compat->presentModeCount = 1;
-            } else {
-               switch (present_mode->presentMode) {
-               case VK_PRESENT_MODE_MAILBOX_KHR:
-               case VK_PRESENT_MODE_FIFO_KHR:
-                  compat->presentModeCount = 2;
-                  break;
-               default:
-                  compat->presentModeCount = 1;
-                  break;
-               }
             }
+
+            uint32_t present_mode_count = 0;
+            result = wsi_wl_surface_get_present_modes(surface, wsi_device,
+                                                      &present_mode_count, NULL);
+            if (result != VK_SUCCESS)
+               return result;
+
+            compat->presentModeCount = present_mode_count;
          }
          break;
       }
@@ -2197,48 +2243,6 @@ wsi_wl_surface_get_formats2(VkIcdSurfaceBase *icd_surface,
    return vk_outarray_status(&out);
 }
 
-static VkResult
-wsi_wl_surface_get_present_modes(VkIcdSurfaceBase *icd_surface,
-                                 struct wsi_device *wsi_device,
-                                 uint32_t* pPresentModeCount,
-                                 VkPresentModeKHR* pPresentModes)
-{
-   VkIcdSurfaceWayland *surface = (VkIcdSurfaceWayland *)icd_surface;
-   struct wsi_wayland *wsi =
-      (struct wsi_wayland *)wsi_device->wsi[VK_ICD_WSI_PLATFORM_WAYLAND];
-
-   struct wsi_wl_display display;
-   if (wsi_wl_display_init(wsi, &display, surface->display, true,
-                           wsi_device->sw, "mesa present modes query"))
-      return VK_ERROR_SURFACE_LOST_KHR;
-
-   VkPresentModeKHR present_modes[3];
-   uint32_t present_modes_count = 0;
-
-   /* The following two modes are always supported */
-   present_modes[present_modes_count++] = VK_PRESENT_MODE_MAILBOX_KHR;
-   present_modes[present_modes_count++] = VK_PRESENT_MODE_FIFO_KHR;
-
-   if (display.tearing_control_manager)
-      present_modes[present_modes_count++] = VK_PRESENT_MODE_IMMEDIATE_KHR;
-
-   assert(present_modes_count <= ARRAY_SIZE(present_modes));
-   wsi_wl_display_finish(&display);
-
-   if (pPresentModes == NULL) {
-      *pPresentModeCount = present_modes_count;
-      return VK_SUCCESS;
-   }
-
-   *pPresentModeCount = MIN2(*pPresentModeCount, present_modes_count);
-   typed_memcpy(pPresentModes, present_modes, *pPresentModeCount);
-
-   if (*pPresentModeCount < present_modes_count)
-      return VK_INCOMPLETE;
-   else
-      return VK_SUCCESS;
-}
-
 static VkResult
 wsi_wl_surface_get_present_rectangles(VkIcdSurfaceBase *surface,
                                       struct wsi_device *wsi_device,
@@ -2605,12 +2609,42 @@ wsi_wl_swapchain_release_images(struct wsi_swapchain *wsi_chain,
    return VK_SUCCESS;
 }
 
+static VkResult
+wsi_wl_swapchain_update_tearing_hint(struct wsi_wl_swapchain *chain)
+{
+   struct wsi_wl_display *dpy = chain->wsi_wl_surface->display;
+
+   if (chain->retired || !dpy->tearing_control_manager)
+      return VK_SUCCESS;
+
+   if (chain->base.present_mode == VK_PRESENT_MODE_IMMEDIATE_KHR) {
+      if (!chain->tearing_control) {
+         chain->tearing_control =
+            wp_tearing_control_manager_v1_get_tearing_control(
+               dpy->tearing_control_manager,
+               chain->wsi_wl_surface->wayland_surface.wrapper);
+         if (!chain->tearing_control)
+            return VK_ERROR_OUT_OF_HOST_MEMORY;
+      }
+      wp_tearing_control_v1_set_presentation_hint(
+         chain->tearing_control,
+         WP_TEARING_CONTROL_V1_PRESENTATION_HINT_ASYNC);
+   } else if (chain->tearing_control) {
+      wp_tearing_control_v1_set_presentation_hint(
+         chain->tearing_control,
+         WP_TEARING_CONTROL_V1_PRESENTATION_HINT_VSYNC);
+   }
+
+   return VK_SUCCESS;
+}
+
 static void
 wsi_wl_swapchain_set_present_mode(struct wsi_swapchain *wsi_chain,
                                   VkPresentModeKHR mode)
 {
    struct wsi_wl_swapchain *chain = (struct wsi_wl_swapchain *)wsi_chain;
    chain->base.present_mode = mode;
+   wsi_wl_swapchain_update_tearing_hint(chain);
 }
 
 static void
@@ -3974,15 +4008,9 @@ wsi_wl_surface_create_swapchain(VkIcdSurfaceBase *icd_surface,
    }
 
    if (present_mode == VK_PRESENT_MODE_IMMEDIATE_KHR) {
-      chain->tearing_control =
-         wp_tearing_control_manager_v1_get_tearing_control(dpy->tearing_control_manager,
-                                                           chain->wsi_wl_surface->wayland_surface.wrapper);
-      if (!chain->tearing_control) {
-         result = VK_ERROR_OUT_OF_HOST_MEMORY;
+      result = wsi_wl_swapchain_update_tearing_hint(chain);
+      if (result != VK_SUCCESS)
          goto fail_free_wl_chain;
-      }
-      wp_tearing_control_v1_set_presentation_hint(chain->tearing_control,
-                                                  WP_TEARING_CONTROL_V1_PRESENTATION_HINT_ASYNC);
    }
 
    if (wsi_wl_use_explicit_sync(dpy, wsi_device)) {
-- 
2.54.0

